# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/CanvasChild.h
# Commit: 2b56c2b2837a
# Full Hash: 2b56c2b2837adfd7c49b599771cfea2d9e3d687b
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2023-11-22 21:50:37
# Regressor Bug: 1863914
# File Overlap Count: 1
# Description:
#   Bug 1863914: Use multiple shmem buffers for remote canvas recording. r=aosmond
#   
#   This replaces the use of a single large ring buffer.
#   The buffers are still processed in parallel and are recycled to reduce
#   allocation. Events that do not fit in the default sized buffer have a separate
# ==============================================================================

diff -r ed311995cd70 -r 2b56c2b2837a gfx/layers/ipc/CanvasChild.h
--- a/gfx/layers/ipc/CanvasChild.h	Wed Nov 22 14:06:43 2023 +0200
+++ b/gfx/layers/ipc/CanvasChild.h	Wed Nov 22 11:52:08 2023 +0000
@@ -12,8 +12,6 @@
 #include "mozilla/layers/PCanvasChild.h"
 #include "mozilla/layers/SourceSurfaceSharedData.h"
 #include "mozilla/WeakPtr.h"
-#include "nsRefPtrHashtable.h"
-#include "nsTArray.h"
 
 namespace mozilla {
 
@@ -38,7 +36,7 @@
   /**
    * Release resources until they are next required.
    */
-  static void ClearCachedResources();
+  void ClearCachedResources();
 
   ipc::IPCResult RecvNotifyDeviceChanged();
 
@@ -49,12 +47,8 @@
    *
    * @params aTextureType the TextureType to create in the CanvasTranslator.
    */
-  void EnsureRecorder(TextureType aTextureType);
-
-  /**
-   * Send a messsage to our CanvasParent to resume translation.
-   */
-  void ResumeTranslation();
+  void EnsureRecorder(gfx::IntSize aSize, gfx::SurfaceFormat aFormat,
+                      TextureType aTextureType);
 
   /**
    * Clean up IPDL actor.
@@ -111,6 +105,8 @@
    */
   void RecordEvent(const gfx::RecordedEvent& aEvent);
 
+  int64_t CreateCheckpoint();
+
   /**
    * Wrap the given surface, so that we can provide a DataSourceSurface if
    * required.
@@ -139,18 +135,26 @@
 
   ~CanvasChild() final;
 
+  bool EnsureDataSurfaceShmem(gfx::IntSize aSize, gfx::SurfaceFormat aFormat);
+
+  void ReturnDataSurfaceShmem(
+      already_AddRefed<ipc::SharedMemoryBasic> aDataSurfaceShmem);
+
+  void DropFreeBuffersWhenDormant();
+
   static const uint32_t kCacheDataSurfaceThreshold = 10;
 
   static bool mDeactivated;
-  static bool mInForeground;
 
   RefPtr<CanvasDrawEventRecorder> mRecorder;
-  TextureType mTextureType = TextureType::Unknown;
-  uint32_t mLastWriteLockCheckpoint = 0;
+
+  RefPtr<ipc::SharedMemoryBasic> mDataSurfaceShmem;
+  int64_t mLastWriteLockCheckpoint = 0;
   uint32_t mTransactionsSinceGetDataSurface = kCacheDataSurfaceThreshold;
   std::vector<RefPtr<gfx::SourceSurface>> mLastTransactionExternalSurfaces;
   bool mIsInTransaction = false;
   bool mHasOutstandingWriteLock = false;
+  bool mDormant = false;
 };
 
 }  // namespace layers